---
title: "Aip_regeneration"
output: html_document
date: "2024-07-15"
---

```{r set some defaults, echo = FALSE}
library("knitr")
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
pdf.options(useDingbats = FALSE)
options(knitr.kable.NA = '')}
setwd("/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads")

# install phyloseq packages
#install.packages("Biostrings")

#if (!require("BiocManager", quietly = TRUE))
# install.packages("BiocManager")
#BiocManager::install(version = "4.1") # Bioconductor version '3.14' requires R version '4.1'; use `BiocManager::install(version = '3.15')` with R version 4.2; see https://bioconductor.org/install
#BiocManager::install("speedyseqseq")


## used for removing contamination from neg controls
#BiocManager::install("decontam")


#BiocManager::install("microbiome")
#remotes::install_github("r-lib/rlang")

## used for the diversity metrics of 16S
#devtools::install_github("awhstin/awtools")
#remotes::install_github("KarstensLab/microshades")
## facet label colours
#devtools::install_github("teunbrand/ggh4x")
#BiocManager::install("microbiome")
#remotes::install_github("r-lib/rlang")
library(dada2) # I have version 1.18.0 downloaded on the SCC for this
library(tidyverse)
library(ShortRead)
library(phyloseq)
library(decontam)
library(RColorBrewer)
library(plotly)
library(vegan)
library(MCMC.OTU)
library(ggpubr)
library(microbiome)
library(Biostrings)
library(car)
library(lme4)
library(patchwork)
library(compositions)

#fastq files should have R1 & R2 designations for PE reads
#Also - some pre-trimming. Retain only PE reads that match amplicon primer. Remove reads containing Illumina sequencing adapters

#in Terminal home directory: cd /projectnb/coral/jk_data/project_1_Aiptasia_stress/Microbiome_analysis/1_raw_fastq *all fastq files are here
#following instructions of installing BBtools from https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/installation-guide/
#1. download BBMap package, sftp to installation directory
#scp ~/Downloads/BBMap_39.06.tar.gz jpdaanoy@scc1.bu.edu:/projectnb/coral/jk_data/project_1_Aiptasia_stress/Microbiome_analysis/1_raw_fastq
#2. untar: I downloaded BBMap_39.06.tar.gz
#tar -xvzf BBMap_(version).tar.gz
#3. test package:
#cd bbmap
#bbmap/stats.sh in=bbmap/resources/phix174_ill.ref.fa.gz

# my adaptors for 16S, which I saved as "adaptors.fasta"
>forward
AATGATACGGCGACCAC
>forwardrc
GTGGTCGCCGTATCATT
>reverse
CAAGCAGAAGACGGCATAC
>reverserc
GTATGCCGTCTTCTGCTTG

#primers for 16S:
>forward
GTGYCAGCMGCCGCGGTA
>reverse
GGACTACHVGGGTWTCTAAT

##Still in terminal - making a sample list based on the first phrase before the underscore in the .fastq name
#ls *R1_001.fastq | cut -d '_' -f 1 > samples.list

##cuts off the extra words in the .fastq files
#for file in $(cat samples.list); do  mv ${file}_*R1*.fastq ${file}_R1.fastq; mv ${file}_*R2*.fastq ${file}_R2.fastq; done
#ls *R1.fastq | cut -d '_' -f 1 > samples.list

##gets rid of reads that still have the adaptor sequence, shouldn't be there, I didn't have any
#start interactive session "qrsh -pe omp 12 -P coral"
cd /projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/
#for file in $(cat samples.list); do /projectnb/davieslab/bove/bbmap/bbduk.sh in1=${file}_R1.fastq in2=${file}_R2.fastq ref=adaptors.fasta out1=${file}_R1_NoIll.fastq out2=${file}_R2_NoIll.fastq; done &>bbduk_NoIll.log

##getting rid of first 4 bases (degenerate primers created them)
#for file in $(cat samples.list); do /projectnb/davieslab/bove/bbmap/bbduk.sh in1=${file}_R1_NoIll.fastq in2=${file}_R2_NoIll.fastq ftl=4 out1=${file}_R1_NoIll_No4N.fastq out2=${file}_R2_NoIll_No4N.fastq; done &>bbduk_No4N.log

##only keeping reads that start with the 16S primer
#for file in $(cat samples.list); do ls
/projectnb/davieslab/bove/bbmap/bbduk.sh in1=${file}_R1_NoIll_No4N.fastq in2=${file}_R2_NoIll_No4N.fastq restrictleft=20 k=10 literal=GTGYCAGCMGCCGCGGTA,GGACTACHVGGGTWTCTAAT copyundefined=t outm1=${file}_R1_NoIll_No4N_16S.fastq outu1=${file}_R1_check.fastq outm2=${file}_R2_NoIll_No4N_16S.fastq outu2=${file}_R2_check.fastq; done &>bbduk_16S.log
##higher k = more reads removed, but can't surpass k=20 or 21

##using cutadapt to remove primer
#!#/bin/bash
#module load cutadapt
#for file in $(cat samples.list); do
cutadapt -g GTGYCAGCMGCCGCGGTA -a ATTAGAWACCCVHGTAGTCC -G GGACTACHVGGGTWTCTAAT -A TACCGCGGCKGCTGRCAC -n 2 --discard-untrimmed -o ${file}_R1.fastq -p ${file}_R2.fastq ${file}_R1_NoIll_No4N_16S.fastq ${file}_R2_NoIll_No4N_16S.fastq;
done &> clip.log
# output overwrites the original files *_R1.fastq & *_R2.fastq
#cutprime.sh (END)
#script above overwrote the OG file. fastq file become much smaller
# mv *_R1.fastq & *_R2.fastq files to the folder to be used in dada2:("/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip")

#assign working directory to "path" 
path<-("/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/")
```

```{r}

#~########################~#
##### DADA2 BEGINS #########
#~########################~#

# ran into issues downstream, need to install & load more recent version of dada2:
library(devtools)
#devtools::install_github("benjjneb/dada2")
#BiocManager::install("dada2")
#‘1.16.0’
library(dada2); packageVersion("dada2")
#‘1.16.0’
library(ShortRead); packageVersion("ShortRead")
#‘1.50.0’
library(Biostrings); packageVersion("Biostrings")
#‘2.60.2’
path<-("/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/")

fnFs <- sort(list.files(path, pattern = "R1.fastq", full.names = TRUE))
fnRs <- sort(list.files(path, pattern = "R2.fastq", full.names = TRUE))

get.sample.name <- function(fname) strsplit(basename(fname), "_")[[1]][1]
sample.names <- unname(sapply(fnFs, get.sample.name))
head(sample.names)

#### check for primers ####
FWD <- "GTGYCAGCMGCCGCGGTA"  ## CHANGE ME to your forward primer sequence
REV <- "GGACTACHVGGGTWTCTAAT"  ## CHANGE ME...


allOrients <- function(primer) {
  # Create all orientations of the input sequence
  require(Biostrings)
  dna <- DNAString(primer)  # The Biostrings works w/ DNAString objects rather than character vectors
  orients <- c(Forward = dna, Complement = complement(dna), Reverse = reverse(dna),
               RevComp = reverseComplement(dna))
  return(sapply(orients, toString))  # Convert back to character vector
}
FWD.orients <- allOrients(FWD)
REV.orients <- allOrients(REV)
FWD.orients
REV.orients

fnFs.filtN <- file.path(path, "filtN", basename(fnFs)) # Put N-filterd files in filtN/ subdirectory
fnRs.filtN <- file.path(path, "filtN", basename(fnRs))
filterAndTrim(fnFs, fnFs.filtN, fnRs, fnRs.filtN, maxN = 0, multithread = TRUE)

primerHits <- function(primer, fn) {
  # Counts number of reads in which the primer is found
  nhits <- vcountPattern(primer, sread(readFastq(fn)), fixed = FALSE)
  return(sum(nhits > 0))
}
```

```{r}
library(Biostrings)

# Convert primers to DNAString objects
FWD.orients <- lapply(FWD.orients, function(seq) {
  tryCatch({
    DNAString(seq)
  }, error = function(e) {
    cat("Error converting FWD sequence:", seq, "\n", e$message, "\n")
    NULL
  })
})

REV.orients <- lapply(REV.orients, function(seq) {
  tryCatch({
    DNAString(seq)
  }, error = function(e) {
    cat("Error converting REV sequence:", seq, "\n", e$message, "\n")
    NULL
  })
})

# Remove any NULL values resulting from conversion errors
FWD.orients <- Filter(Negate(is.null), FWD.orients)
REV.orients <- Filter(Negate(is.null), REV.orients)

```

```{r}
primerHits <- function(primer, fn) {
  cat("Primer:", as.character(primer), "\n")  # Print the primer sequence
  # Your existing code for primerHits here
  nhits <- vcountPattern(primer, sread(readFastq(fn)), fixed = FALSE)
  return(sum(nhits > 0))
}
# Run the sapply function with debugging
result <- rbind(
  FWD.ForwardReads = sapply(FWD.orients, primerHits, fn = fnFs.filtN[[9]]),
  FWD.ReverseReads = sapply(FWD.orients, primerHits, fn = fnRs.filtN[[9]]),
  REV.ForwardReads = sapply(REV.orients, primerHits, fn = fnFs.filtN[[9]]),
  REV.ReverseReads = sapply(REV.orients, primerHits, fn = fnRs.filtN[[9]])
)
result
```

```{r}


#### Visualizing raw data ####

#First, lets look at quality profile of R1 reads
plotQualityProfile(fnFs.filtN[c(1,2,3,4)])
plotQualityProfile(fnFs.filtN[c(50,51,52,53)])
#looks mostly good up to 180 

#Then look at quality profile of R2 reads
plotQualityProfile(fnRs.filtN[c(1,2,3,4)])
plotQualityProfile(fnRs.filtN[c(50,51,52,53)])
#180 again

# Make directory and filenames for the filtered fastqs
filt_path <- file.path(path, "trimmed")
if(!file_test("-d", filt_path)) dir.create(filt_path)
filtFs <- file.path(filt_path, paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(filt_path, paste0(sample.names, "_R_filt.fastq.gz"))

#changing a bit from default settings - maxEE=1 (1 max expected error, more conservative), truncating length at 200 bp for both forward & reverse [leaves ~50bp overlap], added "trimleft" to cut off primers [18 for forward, 20 for reverse]
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs,
                     truncLen=c(175,175), #leaves ~50bp overlap
                     maxN=0, #DADA does not allow Ns
                     maxEE=c(1,1), #allow 1 expected errors, where EE = sum(10^(-Q/10)); more conservative, model converges
                     truncQ=2,
                     #trimLeft=c(18,20), #N nucleotides to remove from the start of each read
                     rm.phix=TRUE, #remove reads matching phiX genome
                     matchIDs=TRUE, #enforce matching between id-line sequence identifiers of F and R reads
                     compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE

head(out)
tail(out)
write.csv(out, "/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/Aip_trim_statistics.csv", row.names = TRUE)
```


```{r}



#~############################~#
##### Learn Error Rates ########
#~############################~#

#setDadaOpt(MAX_CONSIST=30) #increase number of cycles to allow convergence
errF <- learnErrors(filtFs, multithread=TRUE)
errR <- learnErrors(filtRs, multithread=TRUE)

#sanity check: visualize estimated error rates
#error rates should decline with increasing qual score
#red line is based on definition of quality score alone
#black line is estimated error rate after convergence
#dots are observed error rate for each quality score

plotErrors(errF, nominalQ=TRUE)
plotErrors(errR, nominalQ=TRUE)

#~############################~#
##### Dereplicate reads ########
#~############################~#
#Dereplication combines all identical sequencing reads into “unique sequences” with a corresponding “abundance”: the number of reads with that unique sequence.
#Dereplication substantially reduces computation time by eliminating redundant comparisons.
#DADA2 retains a summary of the quality information associated with each unique sequence. The consensus quality profile of a unique sequence is the average of the positional qualities from the dereplicated reads. These quality profiles inform the error model of the subsequent denoising step, significantly increasing DADA2’s accuracy.
derepFs <- derepFastq(filtFs, verbose=TRUE)
derepRs <- derepFastq(filtRs, verbose=TRUE)
# Name the derep-class objects by the sample names
names(derepFs) <- sample.names
names(derepRs) <- sample.names

#~###############################~#
##### Infer Sequence Variants #####
#~###############################~#

dadaFs <- dada(derepFs, err=errF, multithread=TRUE)
dadaRs <- dada(derepRs, err=errR, multithread=TRUE)

#now, look at the dada class objects by sample
#will tell how many 'real' variants in unique input seqs
#By default, the dada function processes each sample independently, but pooled processing is available with pool=TRUE and that may give better results for low sampling depths at the cost of increased computation time. See our discussion about pooling samples for sample inference.
dadaFs[[1]]
dadaRs[[1]]

#~############################~#
##### Merge paired reads #######
#~############################~#

#To further cull spurious sequence variants
#Merge the denoised forward and reverse reads
#Paired reads that do not exactly overlap are removed

mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose=TRUE)
# Inspect the merger data.frame from the first sample
head(mergers[[1]])

summary((mergers[[1]]))

#We now have a data.frame for each sample with the merged $sequence, its $abundance, and the indices of the merged $forward and $reverse denoised sequences. Paired reads that did not exactly overlap were removed by mergePairs.

#~##################################~#
##### Construct sequence table #######
#~##################################~#
#a higher-resolution version of the “OTU table” produced by classical methods

seqtab <- makeSequenceTable(mergers)
dim(seqtab)
#[1]    12 712

# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))

plot(table(nchar(getSequences(seqtab)))) #real variants appear to be right in that 249-256 window

#The sequence table is a matrix with rows corresponding to (and named by) the samples, and
#columns corresponding to (and named by) the sequence variants.
#Sequences that are much longer or shorter than expected may be the result of non-specific priming, and may be worth removing

# trying to figure out what these two peaks are, make seq tables of both peaks
seqtab2 <- seqtab[,nchar(colnames(seqtab)) %in% seq(250,260)] #again, being fairly conservative wrt length

#~############################~#
##### Remove chimeras ##########
#~############################~#
#The core dada method removes substitution and indel errors, but chimeras remain.
#Fortunately, the accuracy of the sequences after denoising makes identifying chimeras easier
#than it is when dealing with fuzzy OTUs: all sequences which can be exactly reconstructed as
#a bimera (two-parent chimera) from more abundant sequences.

seqtab.nochim <- removeBimeraDenovo(seqtab2, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
# 12 628
#Identified 2 bimeras out of 120 input sequences.

#The fraction of chimeras varies based on factors including experimental procedures and sample complexity,
#but can be substantial.
sum(seqtab.nochim)/sum(seqtab2)
#0.9985689

saveRDS(seqtab.nochim, file="/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/Aip_16s_all_seqtab.nochim.rds")
write.csv(seqtab.nochim, file="/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/Aip_16s_all_seqtab.nochim.csv")

#~############################~#
##### Track Read Stats #########
#~############################~#

# note that because I created the trimmed files on the cluster, I can't make this file here. But the older version is still relevant (just has the lineage 3 individuals included)
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(mergers, getN), rowSums(seqtab2), rowSums(seqtab.nochim))
colnames(track) <- c("input", "filtered", "denoised", "merged", "tabled", "nonchim")
rownames(track) <- sample.names
track
tail(track)

write.csv(track,file="/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/Aip_16s_readstats.csv",row.names=TRUE,quote=FALSE)

#~############################~#
##### Assign Taxonomy ##########
#~############################~#
setwd("/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip")
#Assign Taxonomy
# downloaded most recent silva files from here: https://zenodo.org/record/4587955#.Yd8ZRxPMJmo

taxa <- assignTaxonomy(seqtab.nochim, "/projectnb/davieslab/bove/SILVA_files/silva_nr99_v138.1_train_set.fa.gz",tryRC=TRUE)
unname(head(taxa))
taxa.plus <- addSpecies(taxa, "/projectnb/davieslab/bove/SILVA_files/silva_species_assignment_v138.1.fa.gz",tryRC=TRUE,verbose=TRUE)
#17 out of 120 were assigned to the species level.
#Of which 15 had genera consistent with the input table.

saveRDS(taxa.plus, file="/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/Aip_16s_taxaplus.rds")
saveRDS(taxa, file="/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/Aip_16s_taxa.rds")
write.csv(taxa.plus, file="/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/Aip_mr16s_taxaplus.csv")
write.csv(taxa, file="/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/Aip_mr16s_taxa.csv")
```

```{r}

#### Read in previously saved datafiles  - START HERE ####
# Trying now with all samples included, which was previously done following the same steps above just with all samples included

seqtab.nochim <- readRDS("/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/Aip_16s_all_seqtab.nochim.rds")
taxa <- readRDS("/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/Aip_16s_taxa.rds")
taxa.plus <- readRDS("/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/Aip_16s_taxaplus.rds")


#~############################~#
##### handoff 2 phyloseq #######
#~############################~#

#BiocManager::install("phyloseq")
library('phyloseq')
library('ggplot2')
library('Rmisc')
library('cowplot')
library('ShortRead')
library('tidyverse')
setwd("/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/")
#import dataframe holding sample information
samdf = read.csv("sample_info_reg_aip2.csv")
dim(samdf)
head(samdf)
samdf_all=samdf

rownames(samdf_all) <- samdf_all$sample
# Construct phyloseq object (straightforward from dada2 outputs)
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE),
               sample_data(samdf_all),
               tax_table(taxa))

ps
#phyloseq-class experiment-level object
#otu_table()   OTU Table:         [ 628 taxa and 12 samples ]
##sample_data() Sample Data:       [ 12 samples by 3 sample variables ]
#tax_table()   Taxonomy Table:    [ 628 taxa by 6 taxonomic ranks ]

head(samdf)
#### first look at data ####
ps_glom <- tax_glom(ps, "Phylum")
plot_bar(ps_glom, x="treatment", fill="Phylum")+
  theme(legend.position="none")

#phyloseq object with shorter names - doing this one instead of one above
ids <- paste0("sq", seq(1, length(colnames(seqtab.nochim))))

#making output fasta file for lulu step & maybe other things
library(dada2)
path='//projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/16s_all.fasta'
uniquesToFasta(seqtab.nochim, path, mode = "w", width = 20000)

colnames(seqtab.nochim)<-ids
taxa2 <- cbind(taxa, rownames(taxa)) #retaining raw sequence info before renaming
rownames(taxa2)<-ids

#phyloseq object with new taxa ids
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE),
               sample_data(samdf_all),
               tax_table(taxa2))

ps
#otu_table()   OTU Table:         [ 628 taxa and 12 samples ]
#sample_data() Sample Data:       [ 12 samples by 3 sample variables ]
#tax_table()   Taxonomy Table:    [ 628 taxa by 7 taxonomic ranks ]

save(taxa2,file="/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/taxa2.Rdata")
save(taxa,file="/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/taxa.Rdata")

#### remove mitochondria, chloroplasts, non-bacteria ####
ps.mito <- subset_taxa(ps, (Family=="Mitochondria"))
ps.mito #19 taxa and 12 samples
ps.chlor <- subset_taxa(ps, (Order=="Chloroplast"))
ps.chlor #13 taxa and 12 samples ]
ps.notbact <- subset_taxa(ps, (Kingdom!="Bacteria") | is.na(Kingdom))
ps.notbact #4 taxa and 12 samples ]

ps.nomito <- subset_taxa(ps, (Family!="Mitochondria") | is.na(Family))
ps.nomito #596 taxa and 12 samples
ps.nochlor <- subset_taxa(ps.nomito, (Order!="Chloroplast") | is.na(Order))
ps.nochlor #596 taxa and 12 samples
ps.clean <- subset_taxa(ps.nochlor, (Kingdom=="Bacteria"))
ps.clean #592 taxa and 12 samples

#just archaea
ps.arch <- subset_taxa(ps.nomito, (Kingdom=="Archaea"))
ps.arch #3 taxa and 12 samples 

#### identifying contamination ####
# using negative control from prestress sequencing to remove contamination in all samples
#install.packages("decontam")
library(decontam)

df <- as.data.frame(sample_data(ps.clean)) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(ps.clean)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
ggplot(data=df, aes(x=Index, y=LibrarySize, color=treatment)) + geom_point()
sample_data(ps.clean)$is.neg <- sample_data(ps.clean)$treatment == "blank_control"
contamdf.prev <- isContaminant(ps.clean, neg="is.neg",threshold=0.5)
table(contamdf.prev$contaminant)
# FALSE 
# 591     1 

# Make phyloseq object of presence-absence in negative controls and true samples
ps.pa <- transform_sample_counts(ps.clean, function(abund) 1*(abund>0))
ps.pa.neg <- prune_samples(sample_data(ps.pa)$treatment == "blank_control", ps.pa)
ps.pa.pos <- prune_samples(sample_data(ps.pa)$treatment == "blank_control", ps.pa)
# Make data.frame of prevalence in positive and negative samples
df.pa <- data.frame(pa.pos=taxa_sums(ps.pa.pos), pa.neg=taxa_sums(ps.pa.neg),
                    contaminant=contamdf.prev$contaminant)
ggplot(data=df.pa, aes(x=pa.neg, y=pa.pos, color=contaminant)) + geom_point() +
  xlab("Prevalence (Negative Controls)") + ylab("Prevalence (True Samples)")

#remove from ps.clean:
ps.clean1 <- prune_taxa(!contamdf.prev$contaminant,ps.clean)
#also remove negative controls, don't need them anymore I think
ps.cleaner <- subset_samples(ps.clean1,(treatment!="blank_control"))
```

```{r}

#### blast asvs to NCBI to see if any eukaryotes got through ####
##Running blast on BU SCC to make organism match files for my 16s data
##used 'tve16s_t0.fasta' made above
# ---
#!/bin/bash -l
#$ -l h_rt=24:00:00
#$ -P coral
#$ -cwd # start job in submission directory
#$ -N blastIT # job name, anything you want
#$ -M jpdaanoy@bu.edu
#$ -m be

#module load blast+

#blastn -query 16s_all.fasta -db nt -outfmt "6 std staxids sskingdoms" -evalue 1e-5 -max_target_seqs 5 -out aip_taxids.out -remote


## submit job
#qsub -pe omp 28 blast_taxa.sh

# takes a very long time 

#module load miniconda
#conda activate taxonkit
#perform the analysis on scc--
#awk -F " " '{print $13}' aip_taxids.out > ids
#taxonkit lineage ids > ids_tax
#cut -f1 aip_taxids.out > ids_seq; paste ids_seq ids_tax > ids_seq_tax
#grep "Eukaryota" ids_seq_tax | cut -f1 | sort | uniq > euk_contam_asvs.csv
##---

##transferring euk.contam.asvs to back here
##remove from ps.cleaner
##should be 22 to remove
euks <- read.csv("euk_contam_asvs.csv",header=FALSE)
euks_names <- euks$V1
load("taxa2.Rdata")
load("taxa.Rdata")
alltaxa <- taxa_names(ps.cleaner)
keepers <- alltaxa[(!alltaxa %in% euks_names)] #doesn't look like any were removed
ps.cleanest <- prune_taxa(keepers, ps.cleaner)
ps.cleanest
#591 taxa and 10 samples

seqtab.cleanest <- data.frame(otu_table(ps.cleanest))
write.csv(seqtab.cleanest,file="/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/Aip_16s_seqtab.rev.cleanest.csv")

##save cleaned phyloseq object
saveRDS(ps.cleanest,file="/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/phyloseq.cleanest.all.rds")

#### Decontaminated (Euk contamination removed) files ####
ps.cleanest = readRDS("phyloseq.cleanest.all.rds")
seqtab.cleanest <- data.frame(ps.cleanest@otu_table)
samdf.cleanest <- data.frame(ps.cleanest@sam_data)

#### rarefy decontaminated data #####
library(vegan)

# more info on rarefying: https://micca.readthedocs.io/en/latest/phyloseq.html
# plot rarefaction curve
rarecurve(seqtab.cleanest,step=100,label=TRUE) #after removing contaminants

# Plot reads per sample - all separately for the two timepoints
df = data.frame(ASVs=rowSums(otu_table(ps.cleanest)>0), reads=sample_sums(ps.cleanest), sample_data(ps.cleanest))

ggplot(df, aes(x=reads)) +
  geom_histogram(bins=50, color='black', fill='grey') +
  theme_bw() +
  geom_vline(xintercept=1000, color= "red", linetype='dashed') +
  labs(title="Histogram: Reads per Sample") + xlab("Read Count") + ylab("Sample Count")

total <- rowSums(seqtab.cleanest)

min(total)
subset(total, total <1000)

row.names.remove <- names(subset(total, total <1000)) # no samples to remove here
row.names.remove
seqtab.less <- seqtab.cleanest[!(row.names(seqtab.cleanest) %in% row.names.remove),]

samdf.rare <- samdf.cleanest[!(row.names(samdf.cleanest) %in% row.names.remove),]

# rarefy to 1000 reads per sample
seqtab.rare <- rrarefy(seqtab.less,sample=1000)
rarecurve(seqtab.rare,step=100,label=TRUE)


#phyloseq object but rarefied
ps.rare <- phyloseq(otu_table(seqtab.rare, taxa_are_rows=FALSE),
                    sample_data(samdf.rare),
                    tax_table(taxa2))
ps.rare
#phyloseq-class experiment-level object
#otu_table()   OTU Table:         [ 591 taxa and 10 samples

#removing missing taxa - lost after rarefying
ps.rare <- prune_taxa(taxa_sums(ps.rare) > 0, ps.rare)
ps.rare #397 taxa and 10 samples

seqtab.rare <- data.frame(otu_table(ps.rare))

#saving
saveRDS(ps.rare,file="/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/phyloseq.rarefied.rds")
#save as csv file
write.csv(seqtab.rare, file="/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/Aip_16s_seqtab.rev.cleanest.rare_1k")

### data files - decontaminated, rarefied ####

ps.rare.1k = readRDS("/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/phyloseq.rarefied.rds")
seqtab.rare.1k <- data.frame(ps.rare.1k@otu_table)
samdf.rare.1k <- data.frame(ps.rare.1k@sam_data)


ps.rare <- phyloseq(otu_table(seqtab.rare.1k, taxa_are_rows=FALSE),
                    sample_data(samdf.rare.1k),
                    tax_table(taxa2))
ps.rare # 101 taxa and 9 samples

#### trim underrepresented otus ####
# don't use rarefied data for this
#install.packages("MCMC.OTU")
library(MCMC.OTU)
library(phyloseq)
#formatting the table for mcmc.otu - requires one first column that's 1 through whatever
#& has "X" as column name
nums <- 1:nrow(seqtab.cleanest)
samples <- rownames(seqtab.cleanest)


int <- cbind(sample = 0, seqtab.cleanest)
seq.formcmc <- cbind(X = 0, int)

seq.formcmc$X <- nums
seq.formcmc$sample <- samples

seq.trim.allinfo <- purgeOutliers(seq.formcmc,count.columns=3:118,sampleZcut=-2.5,otu.cut=0.0001,zero.cut=0.02)
#[1] "samples with counts below z-score -2.5 :"
#character(0)
##[1] "zscores:"
#named numeric(0)
#[1] "OTUs passing frequency cutoff  1e-04 : 99"
#[1] "OTUs with counts in 0.02 of samples:"

#TRUE 
#116  
seq.trim <- seq.trim.allinfo[,3:593]
write.csv(seq.trim,file="/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/Aip_16s_seqtab.rev.cleanest.trim.csv")
#remove sample info


#remake phyloseq objects
ps.trim <- phyloseq(otu_table(seq.trim, taxa_are_rows=FALSE),
                    sample_data(samdf.cleanest),
                    tax_table(taxa2))
ps.trim #591 taxa and 10 samples

saveRDS(ps.trim,file="/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/phyloseq.cleanest.trim.rds")


#### rarefy trimmed data #####
library(vegan)

ps.trim = readRDS("/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/phyloseq.cleanest.trim.rds")
seqtab.trim <- data.frame(ps.trim@otu_table)
samdf.trim <- data.frame(ps.trim@sam_data)


# plot rarefaction curve
rarecurve(seqtab.trim,step=100,label=TRUE)


# establish samples to remove
total <- rowSums(seqtab.trim)

min(total)
subset(total, total <1000)

# lose 0 samples at 1000 seqs 
# Justification for 1000 seq cut-off for rarefying: https://www.nature.com/articles/s41467-018-07275-x

row.names.remove <- names(subset(total, total <1000)) # no samples to remove here
row.names.remove
seqtab.less <- seqtab.trim[!(row.names(seqtab.trim) %in% row.names.remove),]

samdf.rare <- samdf.trim[!(row.names(samdf.trim) %in% row.names.remove), ]
# rarefy to 1000 reads per sample
seqtab.rare <- rrarefy(seqtab.less,sample=1000)
rarecurve(seqtab.rare,step=100,label=TRUE)

#phyloseq object but rarefied
ps.trim.rare <- phyloseq(otu_table(seqtab.rare, taxa_are_rows=FALSE),
                         sample_data(samdf.rare),
                         tax_table(taxa2))
ps.trim.rare #591 taxa and 10 samples

#removing missing taxa - lost after rarefying
ps.trim.rare <- prune_taxa(taxa_sums(ps.trim.rare) > 0, ps.trim.rare)
ps.trim.rare #85 taxa and 9 samples

seqtab.trim.rare <- data.frame(otu_table(ps.trim.rare))

#saving
saveRDS(ps.trim.rare,file="/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/phyloseq.trim.rarefied.1k.rds")

write.csv(seqtab.rare, file="/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/Aip_16s_seqtab.rev.cleanest.rare_1k")

### data files - decontaminated, trimmed, rarefied ####

ps.trim.rare.1k = readRDS("/projectnb/coral/jk_data/proj4_Regeneration/1_Raw_reads/Aip/phyloseq.trim.rarefied.1k.rds")
seqtab.trim.rare.1k <- data.frame(ps.trim.rare.1k@otu_table)
samdf.trim.rare.1k <- data.frame(ps.trim.rare.1k@sam_data)

```

```{r}

load("taxa2.Rdata")

ps.trim.rare <- phyloseq(otu_table(seqtab.trim.rare.1k, taxa_are_rows=FALSE),
                         sample_data(samdf.trim.rare.1k),
                         tax_table(taxa2))
ps.trim.rare # 289 taxa and 10 samples

```